File: samples/java/DeliveryDispatch/Server/Tracking/src/main/java/systems/zlink/samples/deliverydispatch/server/tracking/handlers/ServerAssertionHandler.java
package systems.zlink.samples.deliverydispatch.server.tracking.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.deliverydispatch.server.configuration.EvidenceStore;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup("tracking")
public final class ServerAssertionHandler
        implements ZLinkRequestHandler<Messages.ServerAssertionReq, Messages.ServerAssertionRes> {
    private final EvidenceStore evidenceStore;

    public ServerAssertionHandler(EvidenceStore evidenceStore) {
        this.evidenceStore = evidenceStore;
    }

    @Override
    public CompletionStage<Messages.ServerAssertionRes> handle(
            Messages.ServerAssertionReq request, ZLinkMessageContext context) {
        return CompletableFuture.completedFuture(
                evidenceStore.assertSequences(
                        request.successfulDeliveryId(), request.reassignedDeliveryId()));
    }
}

File: samples/kotlin/DeliveryDispatch/Server/Tracking/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/tracking/handlers/ServerAssertionHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.tracking.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.DeliveryEvidenceStore
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.ServerAssertionReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.ServerAssertionRes

@ZLinkHandlerGroup("tracking")
class ServerAssertionHandler(private val evidenceStore: DeliveryEvidenceStore) :
    ZLinkSuspendingRequestHandler<ServerAssertionReq, ServerAssertionRes> {
    override suspend fun handle(
        request: ServerAssertionReq,
        context: ZLinkMessageContext,
    ): ServerAssertionRes =
        evidenceStore.assertSequences(request.successfulDeliveryId, request.reassignedDeliveryId)
}
